# Introducción a la Programación para Ciencia de Datos
## Lenguaje de programación R
_Rocío Romero Zaliz_ - rocio@decsai.ugr.es

# Tidyverse
Una colección de paquetes con una gramática, filosofía y estructura similar (https://tidyverse.tidyverse.org)

<i>Wickham H, Averick M, Bryan J, Chang W, McGowan LD, François R, Grolemund G, Hayes A, Henry L, Hester J, Kuhn M, Pedersen TL, Miller E, Bache SM, Müller K, Ooms J, Robinson D, Seidel DP, Spinu V, Takahashi K, Vaughan D, Wilke C, Woo K, Yutani H (2019). “Welcome to the tidyverse.” Journal of Open Source Software, 4(43), 1686. doi:10.21105/joss.01686.</i>

In [1]:
# Carga del paquete
library(tidyverse)

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.5.2     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.0.4     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


## Cheatsheets
https://posit.co/resources/cheatsheets/

In [2]:
class(mtcars)

[1] "data.frame"

In [3]:
mtcars

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

## Pipeline (%>%) - paquete magrittr
Operador que sirve para realizar varias operaciones de forma secuencial sin recurrir a parentesis anidados o a sobrescribir bases de datos.

In [4]:
# Sin magrittr
x <- c(1, 4, 6, 8)
y <- round(mean(sqrt(log(x))), 2)
y

[1] 0.99

In [5]:
# Con magrittr
# library(magrittr)
x <- c(1, 4, 6, 8)
y <- x %>% log() %>% sqrt() %>% mean() %>% round(2)
y

[1] 0.99

In [6]:
# Con magrittr
x <- c(1, 4, 6, 8)
y <- x %>% log %>% sqrt %>% mean %>% round(2)
y

[1] 0.99

In [7]:
log <- 2
log

[1] 2

In [8]:
log(2)

[1] 0.6931472

## Paquete dplyr
dplyr es un paquete de R para manipular, limpiar y resumir datos no estructurados. Facilita y agiliza la exploración y manipulación de datos en R.

In [12]:
# Vamos a trabajar con un conjunto de datos ya creado
# library(dplyr)
starwars %>% print

# A tibble: 87 × 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Luke Sk…    172    77 blond      fair       blue            19   male  mascu…
 2 C-3PO       167    75 NA         gold       yellow         112   none  mascu…
 3 R2-D2        96    32 NA         white, bl… red             33   none  mascu…
 4 Darth V…    202   136 none       white      yellow          41.9 male  mascu…
 5 Leia Or…    150    49 brown      light      brown           19   fema… femin…
 6 Owen La…    178   120 brown, gr… light      blue            52   male  mascu…
 7 Beru Wh…    165    75 brown      light      blue            47   fema… femin…
 8 R5-D4        97    32 NA         white, red red             NA   none  mascu…
 9 Biggs D…    183    84 black      light      brown           24   male  mascu…
10 Obi-Wan…    182    77 auburn, w… fair       blue-gray       57   male  mascu…
# ℹ 77 m

In [19]:
print(class(mtcars))
mtcars$pepe

[1] "data.frame"


NULL

In [ ]:
# Tibble: si accedes a una columna que no existe te avisa...
starwars$pepe

Warning message:
“Unknown or uninitialised column: `pepe`.”


NULL

In [21]:
starwars %>% str

tibble [87 × 14] (S3: tbl_df/tbl/data.frame)
 $ name      : chr [1:87] "Luke Skywalker" "C-3PO" "R2-D2" "Darth Vader" ...
 $ height    : int [1:87] 172 167 96 202 150 178 165 97 183 182 ...
 $ mass      : num [1:87] 77 75 32 136 49 120 75 32 84 77 ...
 $ hair_color: chr [1:87] "blond" NA NA "none" ...
 $ skin_color: chr [1:87] "fair" "gold" "white, blue" "white" ...
 $ eye_color : chr [1:87] "blue" "yellow" "red" "yellow" ...
 $ birth_year: num [1:87] 19 112 33 41.9 19 52 47 NA 24 57 ...
 $ sex       : chr [1:87] "male" "none" "none" "male" ...
 $ gender    : chr [1:87] "masculine" "masculine" "masculine" "masculine" ...
 $ homeworld : chr [1:87] "Tatooine" "Tatooine" "Naboo" "Tatooine" ...
 $ species   : chr [1:87] "Human" "Droid" "Droid" "Human" ...
 $ films     :List of 87
  ..$ : chr [1:5] "A New Hope" "The Empire Strikes Back" "Return of the Jedi" "Revenge of the Sith" ...
  ..$ : chr [1:6] "A New Hope" "The Empire Strikes Back" "Return of the Jedi" "The Phantom Menace" ...
  ..$ 

In [13]:
mtcars$pepe <- starwars$films

ERROR: Error in `$<-.data.frame`(`*tmp*`, pepe, value = list(c("A New Hope", : replacement has 87 rows, data has 32


In [14]:
class(starwars)

[1] "tbl_df"     "tbl"        "data.frame"

### filter()
Selecciona filas en un data frame

In [15]:
mtcars[mtcars$hp > 100,]

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

In [16]:
# Busco a los andorides...
# Fuera del tidyverso
print(starwars[!is.na(starwars$species) & starwars$species == "Droid",])

# A tibble: 6 × 14
  name   height  mass hair_color skin_color  eye_color birth_year sex   gender  
  <chr>   <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>   
1 C-3PO     167    75 NA         gold        yellow           112 none  masculi…
2 R2-D2      96    32 NA         white, blue red               33 none  masculi…
3 R5-D4      97    32 NA         white, red  red               NA none  masculi…
4 IG-88     200   140 none       metal       red               15 none  masculi…
5 R4-P17     96    NA none       silver, red red, blue         NA none  feminine
6 BB8        NA    NA none       none        black             NA none  masculi…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [17]:
# Dentro del tidyverso
starwars %>% filter(species == "Droid") %>% print

# A tibble: 6 × 14
  name   height  mass hair_color skin_color  eye_color birth_year sex   gender  
  <chr>   <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>   
1 C-3PO     167    75 NA         gold        yellow           112 none  masculi…
2 R2-D2      96    32 NA         white, blue red               33 none  masculi…
3 R5-D4      97    32 NA         white, red  red               NA none  masculi…
4 IG-88     200   140 none       metal       red               15 none  masculi…
5 R4-P17     96    NA none       silver, red red, blue         NA none  feminine
6 BB8        NA    NA none       none        black             NA none  masculi…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [18]:
filter(starwars, species == "Droid") %>% print

# A tibble: 6 × 14
  name   height  mass hair_color skin_color  eye_color birth_year sex   gender  
  <chr>   <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>   
1 C-3PO     167    75 NA         gold        yellow           112 none  masculi…
2 R2-D2      96    32 NA         white, blue red               33 none  masculi…
3 R5-D4      97    32 NA         white, red  red               NA none  masculi…
4 IG-88     200   140 none       metal       red               15 none  masculi…
5 R4-P17     96    NA none       silver, red red, blue         NA none  feminine
6 BB8        NA    NA none       none        black             NA none  masculi…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [19]:
starwars %>% filter(species == "Droid") %>% filter(homeworld == "Naboo") %>% print

# A tibble: 1 × 14
  name  height  mass hair_color skin_color  eye_color birth_year sex   gender   
  <chr>  <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>    
1 R2-D2     96    32 NA         white, blue red               33 none  masculine
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [20]:
starwars %>% filter(species == "Droid") %>% filter(height < 100) %>% print

# A tibble: 3 × 14
  name   height  mass hair_color skin_color  eye_color birth_year sex   gender  
  <chr>   <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>   
1 R2-D2      96    32 NA         white, blue red               33 none  masculi…
2 R5-D4      97    32 NA         white, red  red               NA none  masculi…
3 R4-P17     96    NA none       silver, red red, blue         NA none  feminine
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [21]:
starwars %>% filter(species == "Droid") %>%
    filter(height < 100 | is.na(height)) %>% print

# A tibble: 4 × 14
  name   height  mass hair_color skin_color  eye_color birth_year sex   gender  
  <chr>   <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>   
1 R2-D2      96    32 NA         white, blue red               33 none  masculi…
2 R5-D4      97    32 NA         white, red  red               NA none  masculi…
3 R4-P17     96    NA none       silver, red red, blue         NA none  feminine
4 BB8        NA    NA none       none        black             NA none  masculi…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [22]:
starwars %>% filter(species == "Droid") %>%
    filter(height >= 96 & height < 200) %>% print

# A tibble: 4 × 14
  name   height  mass hair_color skin_color  eye_color birth_year sex   gender  
  <chr>   <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>   
1 C-3PO     167    75 NA         gold        yellow           112 none  masculi…
2 R2-D2      96    32 NA         white, blue red               33 none  masculi…
3 R5-D4      97    32 NA         white, red  red               NA none  masculi…
4 R4-P17     96    NA none       silver, red red, blue         NA none  feminine
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [23]:
starwars %>% filter(species == "Droid", height >= 96 & height < 200) %>% print

# A tibble: 4 × 14
  name   height  mass hair_color skin_color  eye_color birth_year sex   gender  
  <chr>   <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>   
1 C-3PO     167    75 NA         gold        yellow           112 none  masculi…
2 R2-D2      96    32 NA         white, blue red               33 none  masculi…
3 R5-D4      97    32 NA         white, red  red               NA none  masculi…
4 R4-P17     96    NA none       silver, red red, blue         NA none  feminine
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [24]:
starwars %>% filter(species == "Droid") %>% 
    filter(height >= 96 & homeworld %in% c("Naboo", "Tatooine")) %>% print

# A tibble: 3 × 14
  name  height  mass hair_color skin_color  eye_color birth_year sex   gender   
  <chr>  <int> <dbl> <chr>      <chr>       <chr>          <dbl> <chr> <chr>    
1 C-3PO    167    75 NA         gold        yellow           112 none  masculine
2 R2-D2     96    32 NA         white, blue red               33 none  masculine
3 R5-D4     97    32 NA         white, red  red               NA none  masculine
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [27]:
print(starwars[10:15,])

# A tibble: 6 × 14
  name      height  mass hair_color skin_color eye_color birth_year sex   gender
  <chr>      <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
1 Obi-Wan …    182    77 auburn, w… fair       blue-gray       57   male  mascu…
2 Anakin S…    188    84 blond      fair       blue            41.9 male  mascu…
3 Wilhuff …    180    NA auburn, g… fair       blue            64   male  mascu…
4 Chewbacca    228   112 brown      unknown    blue           200   male  mascu…
5 Han Solo     180    80 brown      fair       brown           29   male  mascu…
6 Greedo       173    74 NA         green      black           44   male  mascu…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [25]:
starwars %>% slice(10:15) %>% print

# A tibble: 6 × 14
  name      height  mass hair_color skin_color eye_color birth_year sex   gender
  <chr>      <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
1 Obi-Wan …    182    77 auburn, w… fair       blue-gray       57   male  mascu…
2 Anakin S…    188    84 blond      fair       blue            41.9 male  mascu…
3 Wilhuff …    180    NA auburn, g… fair       blue            64   male  mascu…
4 Chewbacca    228   112 brown      unknown    blue           200   male  mascu…
5 Han Solo     180    80 brown      fair       brown           29   male  mascu…
6 Greedo       173    74 NA         green      black           44   male  mascu…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [26]:
starwars %>% slice_max(n=5, height) %>% print # slice_max

# A tibble: 5 × 14
  name      height  mass hair_color skin_color eye_color birth_year sex   gender
  <chr>      <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
1 Yarael P…    264    NA none       white      yellow            NA male  mascu…
2 Tarfful      234   136 brown      brown      blue              NA male  mascu…
3 Lama Su      229    88 none       grey       black             NA male  mascu…
4 Chewbacca    228   112 brown      unknown    blue             200 male  mascu…
5 Roos Tar…    224    82 none       grey       orange            NA male  mascu…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


In [31]:
starwars %>% slice_min(n=5, height) %>% print # slice_min

# A tibble: 6 × 14
  name      height  mass hair_color skin_color eye_color birth_year sex   gender
  <chr>      <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
1 Yoda          66    17 white      green      brown            896 male  mascu…
2 Ratts Ty…     79    15 none       grey, blue unknown           NA male  mascu…
3 Wicket S…     88    20 brown      brown      brown              8 male  mascu…
4 Dud Bolt      94    45 none       blue, grey yellow            NA male  mascu…
5 R2-D2         96    32 NA         white, bl… red               33 none  mascu…
6 R4-P17        96    NA none       silver, r… red, blue         NA none  femin…
# ℹ 5 more variables: homeworld <chr>, species <chr>, films <list>,
#   vehicles <list>, starships <list>


### select()
Permite seleccionar variables (columnas) del data frame

In [32]:
print(starwars[,c("name","homeworld")])

# A tibble: 87 × 2
   name               homeworld
   <chr>              <chr>    
 1 Luke Skywalker     Tatooine 
 2 C-3PO              Tatooine 
 3 R2-D2              Naboo    
 4 Darth Vader        Tatooine 
 5 Leia Organa        Alderaan 
 6 Owen Lars          Tatooine 
 7 Beru Whitesun Lars Tatooine 
 8 R5-D4              Tatooine 
 9 Biggs Darklighter  Tatooine 
10 Obi-Wan Kenobi     Stewjon  
# ℹ 77 more rows


In [33]:
starwars %>% select(name, homeworld) %>% print

# A tibble: 87 × 2
   name               homeworld
   <chr>              <chr>    
 1 Luke Skywalker     Tatooine 
 2 C-3PO              Tatooine 
 3 R2-D2              Naboo    
 4 Darth Vader        Tatooine 
 5 Leia Organa        Alderaan 
 6 Owen Lars          Tatooine 
 7 Beru Whitesun Lars Tatooine 
 8 R5-D4              Tatooine 
 9 Biggs Darklighter  Tatooine 
10 Obi-Wan Kenobi     Stewjon  
# ℹ 77 more rows


In [34]:
starwars %>% select(-name, -homeworld) %>% print

# A tibble: 87 × 12
   height  mass hair_color  skin_color eye_color birth_year sex   gender species
    <int> <dbl> <chr>       <chr>      <chr>          <dbl> <chr> <chr>  <chr>  
 1    172    77 blond       fair       blue            19   male  mascu… Human  
 2    167    75 NA          gold       yellow         112   none  mascu… Droid  
 3     96    32 NA          white, bl… red             33   none  mascu… Droid  
 4    202   136 none        white      yellow          41.9 male  mascu… Human  
 5    150    49 brown       light      brown           19   fema… femin… Human  
 6    178   120 brown, grey light      blue            52   male  mascu… Human  
 7    165    75 brown       light      blue            47   fema… femin… Human  
 8     97    32 NA          white, red red             NA   none  mascu… Droid  
 9    183    84 black       light      brown           24   male  mascu… Human  
10    182    77 auburn, wh… fair       blue-gray       57   male  mascu… Human  
# ℹ 77 m

In [35]:
starwars %>% select(starts_with("s")) %>% head(5) %>% print

# A tibble: 5 × 4
  skin_color  sex    species starships
  <chr>       <chr>  <chr>   <list>   
1 fair        male   Human   <chr [2]>
2 gold        none   Droid   <chr [0]>
3 white, blue none   Droid   <chr [0]>
4 white       male   Human   <chr [1]>
5 light       female Human   <chr [0]>


In [36]:
starwars %>% select(ends_with("es")) %>% head(5) %>% print

# A tibble: 5 × 2
  species vehicles 
  <chr>   <list>   
1 Human   <chr [2]>
2 Droid   <chr [0]>
3 Droid   <chr [0]>
4 Human   <chr [0]>
5 Human   <chr [1]>


In [42]:
starwars %>% select(contains("a")) %>% head(5) %>% print

# A tibble: 5 × 5
  name            mass hair_color birth_year starships
  <chr>          <dbl> <chr>           <dbl> <list>   
1 Luke Skywalker    77 blond            19   <chr [2]>
2 C-3PO             75 NA              112   <chr [0]>
3 R2-D2             32 NA               33   <chr [0]>
4 Darth Vader      136 none             41.9 <chr [1]>
5 Leia Organa       49 brown            19   <chr [0]>


In [40]:
starwars %>% select(name)

name                 
1  Luke Skywalker       
2  C-3PO                
3  R2-D2                
4  Darth Vader          
5  Leia Organa          
6  Owen Lars            
7  Beru Whitesun Lars   
8  R5-D4                
9  Biggs Darklighter    
10 Obi-Wan Kenobi       
11 Anakin Skywalker     
12 Wilhuff Tarkin       
13 Chewbacca            
14 Han Solo             
15 Greedo               
16 Jabba Desilijic Tiure
17 Wedge Antilles       
18 Jek Tono Porkins     
19 Yoda                 
20 Palpatine            
21 Boba Fett            
22 IG-88                
23 Bossk                
24 Lando Calrissian     
25 Lobot                
26 Ackbar               
27 Mon Mothma           
28 Arvel Crynyd         
29 Wicket Systri Warrick
30 Nien Nunb            
⋮  ⋮                    
58 Mas Amedda           
59 Gregar Typho         
60 Cordé                
61 Cliegg Lars          
62 Poggle the Lesser    
63 Luminara Unduli      
64 Barriss Offee        
65 Dormé                
66 Dooku                
67 Bail Prestor Organa  
68 Jango Fett           
69 Zam Wesell           
70 Dexter Jettster      
71 Lama Su              
72 Taun We              
73 Jocasta Nu           
74 R4-P17               
75 Wat Tambor           
76 San Hill             
77 Shaak Ti             
78 Grievous             
79 Tarfful              
80 Raymus Antilles      
81 Sly Moore            
82 Tion Medon           
83 Finn                 
84 Rey                  
85 Poe Dameron          
86 BB8                  
87 Captain Phasma

In [39]:
starwars %>% pull(name)

[1] "Luke Skywalker"        "C-3PO"                 "R2-D2"                
 [4] "Darth Vader"           "Leia Organa"           "Owen Lars"            
 [7] "Beru Whitesun Lars"    "R5-D4"                 "Biggs Darklighter"    
[10] "Obi-Wan Kenobi"        "Anakin Skywalker"      "Wilhuff Tarkin"       
[13] "Chewbacca"             "Han Solo"              "Greedo"               
[16] "Jabba Desilijic Tiure" "Wedge Antilles"        "Jek Tono Porkins"     
[19] "Yoda"                  "Palpatine"             "Boba Fett"            
[22] "IG-88"                 "Bossk"                 "Lando Calrissian"     
[25] "Lobot"                 "Ackbar"                "Mon Mothma"           
[28] "Arvel Crynyd"          "Wicket Systri Warrick" "Nien Nunb"            
[31] "Qui-Gon Jinn"          "Nute Gunray"           "Finis Valorum"        
[34] "Padmé Amidala"         "Jar Jar Binks"         "Roos Tarpals"         
[37] "Rugor Nass"            "Ric Olié"              "Watto"                
[40] "Sebulba"               "Quarsh Panaka"         "Shmi Skywalker"       
[43] "Darth Maul"            "Bib Fortuna"           "Ayla Secura"          
[46] "Ratts Tyerel"          "Dud Bolt"              "Gasgano"              
[49] "Ben Quadinaros"        "Mace Windu"            "Ki-Adi-Mundi"         
[52] "Kit Fisto"             "Eeth Koth"             "Adi Gallia"           
[55] "Saesee Tiin"           "Yarael Poof"           "Plo Koon"             
[58] "Mas Amedda"            "Gregar Typho"          "Cordé"                
[61] "Cliegg Lars"           "Poggle the Lesser"     "Luminara Unduli"      
[64] "Barriss Offee"         "Dormé"                 "Dooku"                
[67] "Bail Prestor Organa"   "Jango Fett"            "Zam Wesell"           
[70] "Dexter Jettster"       "Lama Su"               "Taun We"              
[73] "Jocasta Nu"            "R4-P17"                "Wat Tambor"           
[76] "San Hill"              "Shaak Ti"              "Grievous"             
[79] "Tarfful"               "Raymus Antilles"       "Sly Moore"            
[82] "Tion Medon"            "Finn"                  "Rey"                  
[85] "Poe Dameron"           "BB8"                   "Captain Phasma"

In [43]:
class(starwars %>% pull(name))

[1] "character"

### arrange()
Reordena filas en un data frame

In [46]:
starwars %>% arrange(height) %>% print

# A tibble: 87 × 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yoda         66    17 white      green      brown            896 male  mascu…
 2 Ratts T…     79    15 none       grey, blue unknown           NA male  mascu…
 3 Wicket …     88    20 brown      brown      brown              8 male  mascu…
 4 Dud Bolt     94    45 none       blue, grey yellow            NA male  mascu…
 5 R2-D2        96    32 NA         white, bl… red               33 none  mascu…
 6 R4-P17       96    NA none       silver, r… red, blue         NA none  femin…
 7 R5-D4        97    32 NA         white, red red               NA none  mascu…
 8 Sebulba     112    40 none       grey, red  orange            NA male  mascu…
 9 Gasgano     122    NA none       white, bl… black             NA male  mascu…
10 Watto       137    NA black      blue, grey yellow            NA male  mascu…
# ℹ 77 m

In [47]:
starwars %>% arrange(desc(height)) %>% print

# A tibble: 87 × 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yarael …    264    NA none       white      yellow          NA   male  mascu…
 2 Tarfful     234   136 brown      brown      blue            NA   male  mascu…
 3 Lama Su     229    88 none       grey       black           NA   male  mascu…
 4 Chewbac…    228   112 brown      unknown    blue           200   male  mascu…
 5 Roos Ta…    224    82 none       grey       orange          NA   male  mascu…
 6 Grievous    216   159 none       brown, wh… green, y…       NA   male  mascu…
 7 Taun We     213    NA none       grey       black           NA   fema… femin…
 8 Rugor N…    206    NA none       green      orange          NA   male  mascu…
 9 Tion Me…    206    80 none       grey       black           NA   male  mascu…
10 Darth V…    202   136 none       white      yellow          41.9 male  mascu…
# ℹ 77 m

In [48]:
starwars %>% arrange(height, desc(birth_year)) %>% print

# A tibble: 87 × 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <chr> <chr> 
 1 Yoda         66    17 white      green      brown            896 male  mascu…
 2 Ratts T…     79    15 none       grey, blue unknown           NA male  mascu…
 3 Wicket …     88    20 brown      brown      brown              8 male  mascu…
 4 Dud Bolt     94    45 none       blue, grey yellow            NA male  mascu…
 5 R2-D2        96    32 NA         white, bl… red               33 none  mascu…
 6 R4-P17       96    NA none       silver, r… red, blue         NA none  femin…
 7 R5-D4        97    32 NA         white, red red               NA none  mascu…
 8 Sebulba     112    40 none       grey, red  orange            NA male  mascu…
 9 Gasgano     122    NA none       white, bl… black             NA male  mascu…
10 Watto       137    NA black      blue, grey yellow            NA male  mascu…
# ℹ 77 m

### rename()
Renombra columnas en un data frame

In [49]:
starwars %>% rename(hair = hair_color) %>% print

# A tibble: 87 × 14
   name          height  mass hair  skin_color eye_color birth_year sex   gender
   <chr>          <int> <dbl> <chr> <chr>      <chr>          <dbl> <chr> <chr> 
 1 Luke Skywalk…    172    77 blond fair       blue            19   male  mascu…
 2 C-3PO            167    75 NA    gold       yellow         112   none  mascu…
 3 R2-D2             96    32 NA    white, bl… red             33   none  mascu…
 4 Darth Vader      202   136 none  white      yellow          41.9 male  mascu…
 5 Leia Organa      150    49 brown light      brown           19   fema… femin…
 6 Owen Lars        178   120 brow… light      blue            52   male  mascu…
 7 Beru Whitesu…    165    75 brown light      blue            47   fema… femin…
 8 R5-D4             97    32 NA    white, red red             NA   none  mascu…
 9 Biggs Darkli…    183    84 black light      brown           24   male  mascu…
10 Obi-Wan Keno…    182    77 aubu… fair       blue-gray       57   male  mascu…
# ℹ 77 m

###  mutate()
Crea nueva columnas en un data frame o actualiza las ya existentes

In [ ]:
starwars %>% 
    mutate(height_in = height * 0.393701) %>% 
    select(starts_with("he")) %>% head(5)

height height_in
1 172    67.71657 
2 167    65.74807 
3  96    37.79530 
4 202    79.52760 
5 150    59.05515

In [56]:
starwars %>% select(sex) %>% head(5)

sex   
1 male  
2 none  
3 none  
4 male  
5 female

In [58]:
starwars %>% 
    mutate(sex = as.factor(sex), height = 1) %>% print

# A tibble: 87 × 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <chr>     <dbl> <dbl> <chr>      <chr>      <chr>          <dbl> <fct> <chr> 
 1 Luke Sk…      1    77 blond      fair       blue            19   male  mascu…
 2 C-3PO         1    75 NA         gold       yellow         112   none  mascu…
 3 R2-D2         1    32 NA         white, bl… red             33   none  mascu…
 4 Darth V…      1   136 none       white      yellow          41.9 male  mascu…
 5 Leia Or…      1    49 brown      light      brown           19   fema… femin…
 6 Owen La…      1   120 brown, gr… light      blue            52   male  mascu…
 7 Beru Wh…      1    75 brown      light      blue            47   fema… femin…
 8 R5-D4         1    32 NA         white, red red             NA   none  mascu…
 9 Biggs D…      1    84 black      light      brown           24   male  mascu…
10 Obi-Wan…      1    77 auburn, w… fair       blue-gray       57   male  mascu…
# ℹ 77 m

También existen funciones `mutate_if` y `mutate_at`.

In [59]:
starwars %>% mutate_if(is.character, as.factor) %>% print

# A tibble: 87 × 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <fct>     <int> <dbl> <fct>      <fct>      <fct>          <dbl> <fct> <fct> 
 1 Luke Sk…    172    77 blond      fair       blue            19   male  mascu…
 2 C-3PO       167    75 NA         gold       yellow         112   none  mascu…
 3 R2-D2        96    32 NA         white, bl… red             33   none  mascu…
 4 Darth V…    202   136 none       white      yellow          41.9 male  mascu…
 5 Leia Or…    150    49 brown      light      brown           19   fema… femin…
 6 Owen La…    178   120 brown, gr… light      blue            52   male  mascu…
 7 Beru Wh…    165    75 brown      light      blue            47   fema… femin…
 8 R5-D4        97    32 NA         white, red red             NA   none  mascu…
 9 Biggs D…    183    84 black      light      brown           24   male  mascu…
10 Obi-Wan…    182    77 auburn, w… fair       blue-gray       57   male  mascu…
# ℹ 77 m

In [60]:
starwars %>% mutate_at(c("name","sex"), as.factor) %>% print

# A tibble: 87 × 14
   name     height  mass hair_color skin_color eye_color birth_year sex   gender
   <fct>     <int> <dbl> <chr>      <chr>      <chr>          <dbl> <fct> <chr> 
 1 Luke Sk…    172    77 blond      fair       blue            19   male  mascu…
 2 C-3PO       167    75 NA         gold       yellow         112   none  mascu…
 3 R2-D2        96    32 NA         white, bl… red             33   none  mascu…
 4 Darth V…    202   136 none       white      yellow          41.9 male  mascu…
 5 Leia Or…    150    49 brown      light      brown           19   fema… femin…
 6 Owen La…    178   120 brown, gr… light      blue            52   male  mascu…
 7 Beru Wh…    165    75 brown      light      blue            47   fema… femin…
 8 R5-D4        97    32 NA         white, red red             NA   none  mascu…
 9 Biggs D…    183    84 black      light      brown           24   male  mascu…
10 Obi-Wan…    182    77 auburn, w… fair       blue-gray       57   male  mascu…
# ℹ 77 m

### summarise()/summarize()
Permite colapsar/resumir filas en un data frame

In [61]:
summary(starwars)

     name               height           mass          hair_color       
 Length:87          Min.   : 66.0   Min.   :  15.00   Length:87         
 Class :character   1st Qu.:167.0   1st Qu.:  55.60   Class :character  
 Mode  :character   Median :180.0   Median :  79.00   Mode  :character  
                    Mean   :174.6   Mean   :  97.31                     
                    3rd Qu.:191.0   3rd Qu.:  84.50                     
                    Max.   :264.0   Max.   :1358.00                     
                    NA's   :6       NA's   :28                          
                                                                        
                                                                        
                                                                        
                                                                        
                                                                        
                                                   

In [62]:
starwars$height

[1] 172 167  96 202 150 178 165  97 183 182 188 180 228 180 173 175 170 180  66
[20] 170 183 200 190 177 175 180 150  NA  88 160 193 191 170 185 196 224 206 183
[39] 137 112 183 163 175 180 178  79  94 122 163 188 198 196 171 184 188 264 188
[58] 196 185 157 183 183 170 166 165 193 191 183 168 198 229 213 167  96 193 191
[77] 178 216 234 188 178 206  NA  NA  NA  NA  NA

In [63]:
starwars$height %>% mean(na.rm=TRUE)

[1] 174.6049

In [64]:
starwars %>% summarise(promedio = mean(height, na.rm=TRUE), NN = n(), desv=sd(mass))

promedio NN desv
1 174.6049 87 NA

In [ ]:
starwars %>% filter(species == "Droid") %>% summarise(NN = n()) # SELECT COUNT(*) FROM starwars WHERE species='Droid'

NN
1 6

In [69]:
starwars %>% filter(species == "Droid") %>% count

n
1 6

In [70]:
starwars %>% summarise(desviacion_tipica = sd(height, na.rm=TRUE))

desviacion_tipica
1 34.77416

In [71]:
starwars %>% summarise(max(mass, na.rm=TRUE))

max(mass, na.rm = TRUE)
1 1358

In [72]:
starwars %>% summarise(mean = mean(height, na.rm=TRUE), sd = sd(height, na.rm=TRUE))

mean     sd      
1 174.6049 34.77416

In [74]:
starwars$species %>% unique %>% length

[1] 38

In [75]:
starwars %>% summarise(n(), mean(mass, na.rm = TRUE))

n() mean(mass, na.rm = TRUE)
1 87  97.31186

In [76]:
# Atención que se vienen curvas...
starwars %>%
        group_by(species) %>%
        summarise(n = n(), mass = mean(mass, na.rm = TRUE))

species        n  mass   
1  Aleena          1   15.00
2  Besalisk        1  102.00
3  Cerean          1   82.00
4  Chagrian        1     NaN
5  Clawdite        1   55.00
6  Droid           6   69.75
7  Dug             1   40.00
8  Ewok            1   20.00
9  Geonosian       1   80.00
10 Gungan          3   74.00
11 Human          35   81.31
12 Hutt            1 1358.00
13 Iktotchi        1     NaN
14 Kaleesh         1  159.00
15 Kaminoan        2   88.00
16 Kel Dor         1   80.00
17 Mirialan        2   53.10
18 Mon Calamari    1   83.00
19 Muun            1     NaN
20 Nautolan        1   87.00
21 Neimodian       1   90.00
22 Pau'an          1   80.00
23 Quermian        1     NaN
24 Rodian          1   74.00
25 Skakoan         1   48.00
26 Sullustan       1   68.00
27 Tholothian      1   50.00
28 Togruta         1   57.00
29 Toong           1   65.00
30 Toydarian       1     NaN
31 Trandoshan      1  113.00
32 Twi'lek         2   55.00
33 Vulptereen      1   45.00
34 Wookiee         2  124.00
35 Xexto           1     NaN
36 Yoda's species  1   17.00
37 Zabrak          2   80.00
38 NA              4   81.00

In [78]:
starwars %>% 
    group_by(species) %>% 
    summarise(n = n(), mass = mean(mass, na.rm = TRUE)) %>% 
    filter(n > 2, mass > 50)

species n  mass 
1 Droid    6 69.75
2 Gungan   3 74.00
3 Human   35 81.31
4 NA       4 81.00

Más funciones útiles para usar con sumarise():
* Center: mean(), median()
* Spread: sd(), IQR(), mad()
* Range: min(), max(), quantile()
* Position: first(), last(), nth()
* Count: n(), n_distinct()
* Logical: any(), all()

In [93]:
starwars %>% summarise_if(is.numeric, mean, na.rm = TRUE)

height   mass     birth_year
1 174.6049 97.31186 87.56512

In [94]:
starwars %>% summarise_if(is.numeric, list(prom=mean,sd=sd), na.rm = TRUE)

height_prom mass_prom birth_year_prom height_sd mass_sd  birth_year_sd
1 174.6049    97.31186  87.56512        34.77416  169.4572 154.6914

In [96]:
starwars %>% 
    summarise_at(vars(height,mass), mean, na.rm = TRUE)

height   mass    
1 174.6049 97.31186

In [99]:
starwars %>% 
    summarise_at(vars(height,mass), list(mean,sd), na.rm = TRUE)

height_fn1 mass_fn1 height_fn2 mass_fn2
1 174.6049   97.31186 34.77416   169.4572

In [83]:
starwars %>% select(height, mass, birth_year) %>%
    summarise_all(list(minimo = min, maximo = max), na.rm = TRUE)

height_minimo mass_minimo birth_year_minimo height_maximo mass_maximo
1 66            15          8                 264           1358       
  birth_year_maximo
1 896

### across()
Se utiliza para aplicar una operación a varias columnas de un data frame de manera simultánea. Esta es una versión más moderna y más legible.

In [122]:
starwars %>% 
  group_by(species) %>% 
  filter(n() > 1) %>% # ¡Ojo! usad parentesis si no os lía con la función n()
  summarise(across(c(sex, gender, homeworld), n_distinct))

species  sex gender homeworld
1 Droid    1   2       3       
2 Gungan   1   1       1       
3 Human    2   2      15       
4 Kaminoan 2   2       1       
5 Mirialan 1   1       1       
6 Twi'lek  2   2       1       
7 Wookiee  1   1       1       
8 Zabrak   1   1       2       
9 NA       1   1       3

In [ ]:
starwars %>% 
  group_by(species) %>% 
  filter(n() > 1) %>% 
  summarise(across(c(sex, gender, homeworld), list(n_distinct,length))) # WTF? length vs. n

species  sex_1 sex_2 gender_1 gender_2 homeworld_1 homeworld_2
1 Droid    1      6    2         6        3           6         
2 Gungan   1      3    1         3        1           3         
3 Human    2     35    2        35       15          35         
4 Kaminoan 2      2    2         2        1           2         
5 Mirialan 1      2    1         2        1           2         
6 Twi'lek  2      2    2         2        1           2         
7 Wookiee  1      2    1         2        1           2         
8 Zabrak   1      2    1         2        2           2         
9 NA       1      4    1         4        3           4

In [127]:
starwars %>% n_distinct

[1] 87

In [128]:
starwars %>% n

ERROR: Error in n(.): el argumento no fue usado (.)


In [129]:
starwars %>% length

[1] 14

In [130]:
dim(starwars)

[1] 87 14

## Extra

In [131]:
df <- data.frame(period=c("Q1_y2019","Q2_y2019", "Q3_y2019","Q4_y2019"), revenue=c(23,24,27,29))
df

period   revenue
1 Q1_y2019 23     
2 Q2_y2019 24     
3 Q3_y2019 27     
4 Q4_y2019 29

In [ ]:
df %>% separate(period, c("Quarter","Year"), sep="_y") # El contario de separate es unite

Quarter Year revenue
1 Q1      2019 23     
2 Q2      2019 24     
3 Q3      2019 27     
4 Q4      2019 29

In [139]:
ndf <- df %>% extract(period, c("Quarter","Year"), "Q(.*)_y(.*)") # Expresiones regulares... más adelante veremos esto
ndf

Quarter Year revenue
1 1       2019 23     
2 2       2019 24     
3 3       2019 27     
4 4       2019 29

In [152]:
df <- tibble(
  a = c(1, 5, 3),
  b = c(4, 2, 6),
  c = c(7, 3, 1)
)

# Sin rowwise... 
df %>% 
    mutate(max_val = max(a, b, c))  # Devuelve el mismo valor para todas las filas


a b c max_val
1 1 4 7 7      
2 5 2 3 7      
3 3 6 1 7

In [151]:
# Con rowwise: calcula el máximo por fila
df %>%
  rowwise() %>%
  mutate(max_val = max(a, b, c))


a b c max_val
1 1 4 7 7      
2 5 2 3 5      
3 3 6 1 6

### Ejercicios Tidyverse
1) Utiliza el dataset de `mtcars` y el paquete `tidyverse` para:
    * Mostrar las 5 primeras filas del dataset
    * Convertir la variables "cyl", "gear" y "carb" en factores y las variables "vs" y "am" en lógicas, y mostrar la estructura del dataset (usar este dataset transformado de aqui en adelante)
    * Mostrar solo los coches con una potencia ("hp") mayor a 100
    * Seleccionar solo las columnas "mpg", "cyl", "hp" y "qsec" del dataset
    * Calcular la cantidad total de coches para cada valor único en la columna "cyl" (número de cilindros)
    * Encontrar el modelo de coche con la mayor potencia ("hp") y mostrar su información completa
    * Calcular el promedio de potencia de los coches con 8 cilindros
2) Supongamos que tienes un data frame llamado `lego_sets` que contiene información sobre diferentes conjuntos de Lego, incluyendo el nombre del conjunto, el número de piezas, el tema y el año de lanzamiento:
    * Filtra y muestra solo los conjuntos de Lego lanzados en el año 2020
    * Encuentra y muestra el nombre y el número de piezas del conjunto de Lego más grande
    * Calcula la cantidad total de piezas para cada tema y muéstralos en orden descendente por número de piezas
    * Calcula cuántos conjuntos de Lego se lanzaron en cada año y muéstralo ordenado por año de forma ascendente
    * Encuentra los 3 temas más populares (con más conjuntos) y muestra el número de conjuntos y el número total de piezas para cada uno de ellos

<pre>
# Generar datos ficticios
nombres <- paste("Conjunto", 1:80)
piezas <- sample(50:1000, 80, replace = TRUE)
temas <- sample(c("Ciudad", "Espacio", "Arquitectura", "Granja", "Dinosaurios", "Aventuras", "Piratas"), 80, replace = TRUE)
años <- sample(2000:2023, 80, replace = TRUE)

# Crear el data frame
lego_sets <- data.frame(
  Set_Name = nombres,
  Piece_Count = piezas,
  Theme = temas,
  Year = años
)
</pre>
3) Dado un data frame llamado `bebidas` que contiene información sobre la edad, género, tipo de bebida y cantidad de copas consumidas por un grupo de personas, filtra y muestra únicamente a las mujeres mayores de 20 años. Para este suconjunto calcula la media, el máximo, la cantidad de copas totales y la desviación estándar de la edad para cada combinación de tipo de bebida. Finalemnte, agrega una nueva columna que indique cuántas personas se encuentran en cada grupo de tipo de bebida.

<pre>
# Generar datos ficticios
num_elementos <- 100
edades <- sample(18:70, num_elementos, replace = TRUE)
sexos <- sample(c("Hombre", "Mujer"), num_elementos, replace = TRUE)
tipos_bebida <- sample(c("Cerveza", "Vino", "Refresco", "Cóctel", "Agua"), num_elementos, replace = TRUE)
cantidad_copas <- sample(1:5, num_elementos, replace = TRUE)

# Crear el data frame de bebidas
bebidas <- data.frame(
  Edad = edades,
  Sexo = sexos,
  Bebida = tipos_bebida,
  Copas = cantidad_copas
)
</pre>
4) Dado un data frame llamado `peliculas`, indique los 2 géneros con mayor puntuación media por país pero solo de peliculas para mayores de 13 o superiores, que incluya el número de películas en ese género, su beneficio medio (ganancia-presupuesto), la desviación estándar de la puntuación y la puntuación media. Ordenar los resultados por puntuación media por país de forma descendente.

<pre>
# Crear un data frame con 150 filas
num_filas <- 150

# Generar datos ficticios
set.seed(123)  # Para reproducibilidad
paises <- sample(c("EE. UU.", "Reino Unido", "Francia", "España", "Italia"), num_filas, replace = TRUE)
nombres <- paste("Película", 1:num_filas)
años <- sample(1980:2023, num_filas, replace = TRUE)
puntuaciones <- round(runif(num_filas, 1, 10), 1)
tematicas <- sample(c("Acción", "Drama", "Comedia", "Ciencia Ficción", "Animación"), num_filas, replace = TRUE)
directores <- paste("Director", 1:num_filas)
companias <- sample(c("Warner Bros.", "Universal Pictures", "Disney", "Sony Pictures", "Paramount Pictures"), num_filas, replace = TRUE)
presupuestos <- round(runif(num_filas, 1000000, 50000000), 2)
ganancias <- round(presupuestos * runif(num_filas, 0.5, 2.5), 2)
ratings <- sample(c("G", "PG", "PG-13", "R", "NC-17"), num_filas, replace = TRUE)  # Columna de rating ficticio

# Crear el data frame
peliculas <- data.frame(
  País = paises,
  Nombre = nombres,
  Año = años,
  Puntuación = puntuaciones,
  Temática = tematicas,
  Director = directores,
  Compañía = companias,
  Presupuesto = presupuestos,
  Ganancia = ganancias,
  Rating = ratings
)
</pre>
5) Dado un conjunto de datos llamado notas que contiene información sobre las notas de los estudiantes en varias asignaturas, agrega una columna con la nota promedio de cada estudiante. Previamente reemplaza los valores faltantes con 0 (HINT: usa replace_na).

<pre>
# Generar datos ficticios
estudiantes <- paste("Estudiante", 1:50)
notas_matematicas <- sample(c(NA, 5, 6, 7, 8, 9, 10), 50, replace = TRUE)
notas_historia <- sample(c(NA, 4, 5, 6, 7, 8), 50, replace = TRUE)
notas_ciencias <- sample(c(NA, 6, 7, 8, 9, 10), 50, replace = TRUE)

# Crear el data frame
notas <- data.frame(
  Estudiante = estudiantes,
  Nota_Matemáticas = notas_matematicas,
  Nota_Historia = notas_historia,
  Nota_Ciencias = notas_ciencias
)
</pre>